feat(path)!: consolidated path libraries

hacdias opened the pull request:
This is the long-awaited consolidation of the multiple path libraries we've had, which closes #198. For more information, see the changelog in the PR.

Kubo PR: https://github.com/ipfs/kubo/pull/10063.

Jorropo commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
-func (p Path) Segments() []string {
-	cleaned := path.Clean(string(p))
-	segments := strings.Split(cleaned, "/")
+// ImmutablePath is a [Path] which is guaranteed to return "false" to [Mutable].
+type ImmutablePath struct {
+	Path
+}
```
```suggestion
var _ Path = ImmutablePath{}

// ImmutablePath is a [Path] which is guaranteed to return "false" to [Path.Mutable].
type ImmutablePath struct {
	path Path
}

func (ip ImmutablePath) String() string {
 return ip.path.String()
}

func (ip ImmutablePath) Namespace() Namespace {
 return ip.path.Namespace()
}

func (ip ImmutablePath) Mutable() bool {
 return false
}

func (ip ImmutablePath) Root() cid.Cid {
 return ip.path.Root()
}

func (ip ImmutablePath) Segments() []string {
 return ip.path.Segments()
}
```
Prevents consumers from manually wrapping as an immutable path without going through `NewImmutablePath`.

Jorropo commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
+	namespace string
+}
+
+func (p *path) String() string {
```
```suggestion
func (p path) String() string {
```
This is never mutated so no need to use a pointer.

Jorropo commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
-// String converts a path to string.
-func (p Path) String() string {
-	return string(p)
+func (p *path) Mutable() bool {
```
```suggestion
func (p path) Mutable() bool {
```

Jorropo commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
-func (p Path) IsJustAKey() bool {
-	parts := p.Segments()
-	return len(parts) == 2 && (parts[0] == "ipfs" || parts[0] == "ipld")
+func (p *path) Root() cid.Cid {
```
```suggestion
func (p path) Root() cid.Cid {
```

Jorropo commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
-// segment, separately. If there is no more to pop (the path is just a key),
-// the original path is returned.
-func (p Path) PopLastSegment() (Path, string, error) {
+func (p *path) Segments() []string {
```
```suggestion
func (p path) Segments() []string {
```

Jorropo commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
+type path struct {
+	str       string
+	root      cid.Cid
+	namespace string
```
```suggestion
	namespace Namespace
```

Jorropo commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
+	return strings.Split(str, "/")
+}
+
+type resolvedPath struct {
+	path
+	cid       cid.Cid
+	remainder string
+}
+
+func (p *resolvedPath) Cid() cid.Cid {
+	return p.cid
+}
+
+func (p *resolvedPath) Remainder() string {
+	return p.remainder
+}
+
+// NewIPFSPath returns a new "/ipfs" path with the provided CID.
+func NewIPFSPath(cid cid.Cid) ResolvedPath {
+	return &resolvedPath{
+		path: path{
+			str:       fmt.Sprintf("/%s/%s", IPFSNamespace, cid.String()),
+			root:      cid,
+			namespace: IPFSNamespace,
+		},
+		cid:       cid,
+		remainder: "",
 	}
+}
```
```suggestion
type resolvedPath struct {
	path
}

func (p resolvedPath) Remainder() string {
	return ""
}

// NewIPFSPath returns a new "/ipfs" path with the provided CID.
func NewIPFSPath(cid cid.Cid) ResolvedPath {
	return resolvedPath{
		path: path{
			str:       fmt.Sprintf("/%s/%s", IPFSNamespace, cid.String()),
			root:      cid,
			namespace: IPFSNamespace,
		},
		cid:       cid,
		remainder: "",
	}
}
```
you are never returning something different from the underlying path, and the remainder is always empty so you can hardcode that

Jorropo reviewed (changes requested):
I shouldn't look at the implementation before the API so the code review is half done.

I don't understand what is the difference between `ResolvedPath` and `ImmutablePath` ?

Jorropo commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
+	remainder string
+}
+
+func (p *resolvedPath) Cid() cid.Cid {
```
```suggestion
func (p resolvedPath) Cid() cid.Cid {
```
no mutation here too (same for other methods)

aschmahmann commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
+// function simply calls [NewPath] internally with the segments concatenated
+// using a forward slash "/" as separator.
+func NewPathFromSegments(segments ...string) (Path, error) {
+	if len(segments) > 1 {
+		if segments[0] == "" {
+			segments = segments[1:]
+		}
 	}
```
Why are we doing this? It seems like either we're relying on `NewPath` to do things like `gopath.Clean` and erroring on paths without enough segments, or we're not. Will removing this result in problems?

hacdias replied:
Honestly, I'm not sure why I added this. I removed it. I have to add more tests.

aschmahmann commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
+//  2. If the path has a valid CID root but does not have a namespace, the /ipfs
+//     namespace is automatically added.
+//
+// This function returns an error when the given string is not a valid path.
```
We should explain here (or on the interface) what a valid path is. Can it have `////` in it? How about `\n`, or `NUL` or other weird characters?

Maybe the answer is that the interface is more flexible and the structs are less, but if so that should be documented as well.

hacdias replied:
@aschmahmann I just updated the description to mention that the path is sanitized with `gopath.Clean`, but final trailing slash is preserved.

aschmahmann commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
-// FromCid safely converts a cid.Cid type to a Path type.
-func FromCid(c cid.Cid) Path {
-	return Path("/ipfs/" + c.String())
+// ResolvedPath is a [Path] which was resolved to the last resolvable node.
+type ResolvedPath interface {
```
Maybe out of scope for this PR, but hopefully not. I'd like to better document what "last resolvable node" means.

For example:
1. A path that's been fully resolved down to its final IPLD block (i.e. the remainder when traversed does not cross any block boundaries)
2. A path that was resolved as far as the implementation could (e.g. it resolved DNSLink but failed at the IPNS resolution, it resolved a dag-cbor part of the path but couldn't understand a dag-jose part of the path, it couldn't find/access some block in the path, etc.)

It seems like this tries to imply the former, although if true I wonder how much this resolution is particularly needed in comparison to others like:
- I started with `/ipns/` and I want the immutable path
- I want the root CIDs of all the "entities" being traversed
- I want all of the ipns resolution information AND the root CIDs of the entities being traversed

Understanding where/why this is used might help us come up with something better here (and need fewer path wrappers or extra lookups in places like the gateway code).

hacdias replied:
@aschmahmann Indeed, and I do also think that the current meaning is **2**. I checked and the only places across both Boxo and Kubo where it is used is:

- `gateway.ContentPathMetadata`, to represent the `LastSegment`, which I think we only do because it's the type used by the following two:
  - `kubo/client/rpc.ResolvePath` 
  - `kubo/coreapi.NewResolvedPath` 

I think we have the opportunity here to get rid of `ResolvedPath`. From what I can see, a path, in its generic form, is just `/{namespace}/{root}[/path]`, where:

- `namespace` can be `ipfs`, `ipns` or `ipld`.
- `root` can be a CID for all, can be a `ipns.Name` for `ipns` and a FQDN for `ipns` (DNSLink).

The key thing to note is that a `ResolvedPath` `Cid` is not the same as `Root`. `Cid` returns the CID of the last resolvable block (see `path/resolver.Resolver`).


cc @Jorropo because you also had questions about ResolvedPath vs. ImmutablePath

hacdias replied:
See [0a71844](https://github.com/ipfs/boxo/pull/334/commits/0a718449a6f1a3a8f8646783e07aa028dbddecd0), where I removed `ResolvedPath` and use `ImmutablePath`. I think this seems like the way to go.

aschmahmann reviewed (commented):
Thank you for starting this 🙏. The current state of the path libraries is .... too much

hacdias commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
+}
 
-	return newPath, segs[len(segs)-1], nil
+// NewIPNSPath returns a new "/ipns" path with the provided CID.
+func NewIPNSPath(cid cid.Cid) Path {
```
@aschmahmann oh, yeah. I'm gonna change this to `ipns.Name`. I just had started this with cid and rebased it but the IPNS PR has landed 😄

hacdias replied:
I removed IPNS and DNSLink related things for now. They can still be created manually, but we slightly limit the scope here.

hacdias replied:
@lidel I added that. I think it can still be useful, as you can just `p := name.AsPath()`, instead of doing `p, err := path.NewPath(ipns.NamespacePrefix + name.String())` and then checking for the error.

hacdias commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
+// present in the given string. The rules are:
+//
+//  1. If the path has a single component (no slashes) ans it is a valid CID,
+//     an /ipfs path is returned. If the CID is encoded with the Libp2pKey codec,
+//     then a /ipns path is returned.
```
Ok. So this is the behaviour of the current code in main. I am updating that as I agree this is a bit of a mess. This has been changed in [e688ad9](https://github.com/ipfs/boxo/pull/334/commits/e688ad9a343083239029d59652ea31c92b923063).  I had to make some updates in the namesys package to keep accepting legacy DNSLink records.

hacdias replied:
I do agree with you. I think we should remove all the implied namespaces, including for IPFS. I think that will break some other things, at least within Boxo, but we can patch them where needed. From what I can see, it will only affect the processing of DNSLink entries.

hacdias replied:
I pushed 4ba0583b76f365119ddc2e46c22a004a2a81192f, which removes all assumptions.  I do think that's the best way forward.

hacdias commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
+	// change. Note that resolved mutable paths can be immutable.
+	Mutable() bool
+
+	// Root returns the [cid.Cid] of the root object of the path. Root can return
+	// [cid.Undef] for Mutable IPNS paths that use [DNSLink].
+	//
+	// [DNSLink]: https://dnslink.dev/
+	Root() cid.Cid
```
I do agree that we should do this, but we have to be careful. As far as I can see, the only place where the current code in Boxo calls `.Root` is for a `ResolvedPath` and not for the `ImmutablePath`. But this probably also connects with the  conversation below: what is a resolved path? https://github.com/ipfs/boxo/pull/334#discussion_r1290656117

hacdias commented on path/path.go:
```diff
@@ -3,188 +3,301 @@ package path
+
+	// Segments returns the different elements ofd a path, which are delimited
+	// by a forward slash ("/"). The leading slash must be ignored, that is, no
+	// segment should be empty.
```
Could you check the comment update I just made? Another suggestion I would make is to simply remove this. The majority of places where we call `Segments` is to check the namespace. That can now be replaced by `Namespace() == path.Namespace`.

hacdias commented:
This is ready for review. You can see Kubo in https://github.com/ipfs/kubo/pull/10063.

aschmahmann commented on path/error.go:
```diff
@@ -1,25 +1,32 @@
 func (e ErrInvalidPath) Is(err error) bool {
 	switch err.(type) {
-	case ErrInvalidPath:
+	case ErrInvalidPath, *ErrInvalidPath:
```
Is this necessary? Seems like this isn't unique to this case and happens everywhere there is an `errors.Is` check, but I don't think doing this double check is that standard. There might be linters that check for this error as IIRC I've seen Goland (Jetbrains IDE) complain if you use the wrong error type (with or without a pointer) when calling `.Is`

aschmahmann commented on path/path.go:
```diff
@@ -3,187 +3,240 @@ package path
-// SplitList splits strings usings /
-func SplitList(pth string) []string {
-	return strings.Split(pth, "/")
+// NewIPLDPath returns a new "/ipld" path with the provided CID.
+func NewIPLDPath(cid cid.Cid) ImmutablePath {
```
I'm not sure I understand the point of this function. This probably means we need more description around what the IPFS and IPLD namespaces are for.

As I understand it:
- `/ipfs` -> UnixFS pathing semantics with fallback to IPLD data model pathing semantics. Don't recall expectations around empty path segments or awkward characters (e.g. `NUL`).
- `/ipld` -> Just IPLD data model pathing. AFAIK this is form of path has been rarely used so far. Lack of definition for where ADLs, etc. fit in has been rough here. It's also not really defined anywhere so it's in a bit of limbo.

Are there supposed to be semantics for what happens at the end of the path with `/ipfs` and `/ipld` for this interface? Are we defining a resource or just how to get to the terminal element and then what happens with that resource is up to the consumer? From what I can tell it looks like the latter, although at some point we may also need one for the former (e.g. that could include things like byte ranges, IPLD ADLs, etc.).

I understand this mess is already here and we may not want to broach this particular mess right now, but even if not it's probably worth some documentation to help clarify the state of things and point to a GitHub issue for further discussion if people run into problems.

hacdias replied:
> I'm not sure I understand the point of this function.

It just creates a new path under the `/ipld` namespace with that CID, so `/ipld/{cid}`. I just kept it because we're using it in other parts of the code.

> From what I can tell it looks like the latter

Yes.

> I understand this mess is already here and we may not want to broach this particular mess right now, but even if not it's probably worth some documentation to help clarify the state of things and point to a GitHub issue for further discussion if people run into problems.

Yeah, I'm not sure if we should tackle this now. I think it also relates to your other comment about [pathing through empty names](https://github.com/ipfs/boxo/pull/334#discussion_r1315988175) (I cannot reply to that comment, I don't know why).

> https://github.com/ipfs/boxo/pull/334#discussion_r1315988175

I think that is a valid concern. My problem is this is the following: before, we were not "cleaning" the path when creating a new path, or when validating it. However, when we split it into segments we were: https://github.com/ipfs/boxo/blob/574df963653c940ea19b1b8a0d37cd30360c22e9/path/path.go#L36 - which I would argue would give an inconsistent result. I don't think that it is very likely that we will want to path through empty namespaces, or namespaces that get collapsed by `gopath.Clean`, such as `.` and `..`. If we do, we will definitely need to update this and maybe have a different type for web paths (cc @lidel as there is some issue for specs for that), unixfs paths, and whatnot. I would argue that at the moment there's no real use case for the things that get collapsed/cleaned by `gopath.Clean`.

hacdias replied:
Removed, and `FromCid` restored.

aschmahmann commented on path/path.go:
```diff
@@ -3,187 +3,240 @@ package path
+	Cid() cid.Cid
 
-	return newPath, segs[len(segs)-1], nil
+	// Remainder returns the unresolved parts of the path.
+	Remainder() string
```
> is what's left to be resolved, or cannot be resolved

But this indicates that it's just any path components beyond the CID, doesn't it? https://github.com/ipfs/boxo/pull/334/files#diff-a44166f2b1bbfa22a31664836b3206fbcfb31600bac5a85b0837aa0644131951R142-R147

> I don't understand this question. We're not returning a []string.

We're returning `[]string`  in places like `resolver.ResolveToLastNode` https://github.com/ipfs/boxo/pull/334/files#diff-0731f9658c3343bd65daf65521d04f5064c46ac26dd0d9bce182db016a8f0ba4R82. That's not a change in this PR, but it's somewhere that returning a path with a `Remainder` might be plausible depending on how we've defined remainder.

---

An area where I'm not sure if there are ramifications here, but maybe you can help me understand since you've been going through all of this now, is if it matters now that we:
- Used to have: ResolvedPath has Root CID, Path, Terminal CID, remainder
- Now have: ImmutablePath has Root CID, Path (remainder == path)

This loss of remainder information seems like it could be problematic. For example, f it was meant to convey "hey I tried resolving through this block but there are no more links, just other IPLD data model fields" then that information would need to be recomputed or specified as required in other interface contracts.

hacdias replied:
@aschmahmann there are surprisingly little places where we use the concept of `.Remainder()` in both Boxo and Kubo. Boxo mainly uses it in the [handler codec](https://github.com/ipfs/boxo/blob/32140848b581e44bc044a66af149814732174915/gateway/handler_codec.go#L84-L92) to ensure there's nothing going inside the path, and Kubo only uses it as a result of [`ResolvePath`
](https://github.com/search?q=repo%3Aipfs%2Fkubo+Remainder%28%29&type=code). I have a suggestion: remove `Remainder` from `ImmutablePath`, and change the `ResolvePath` signature to return `ImmutablePath, []string, error`, matching `ResolveToLastNode`. What do you think? This way, we remove the ambiguous value in the path type and keep it for resolution.

hacdias replied:
@aschmahmann I made the changes mentioned above:

- Boxo: https://github.com/ipfs/boxo/pull/334/commits/4276346cc566a4033b069d08af9d0ea7323119fc
- Kubo: https://github.com/ipfs/kubo/pull/10063/commits/f5b41755bfe4033167f20c6a787802849a14fed5

Let me know what you think. I think this makes the APIs more consistent, and removes the unclear `.Remainder`. Now the remainder is only returned by resolvers and resolving functions, making it clearer what it means and what it should be used for.

hacdias commented:
@aschmahmann @lidel I want to ping you for a re-review. I've already addressed the feedback, tests are passing here and [in Kubo](https://github.com/ipfs/kubo/pull/10063). This PR is almost turning 4 months old and I really don't want that to happen.

lidel reviewed (changes requested):
Some feedback inline, mostly about making interface more clar and avoiding noice caused by legacy decisions in Kubo.

To avoid making waves, I think this should ship in Kubo 0.24, and we should wait with merge until 0.23 ships.

hacdias commented on CHANGELOG.md:
```diff
@@ -24,6 +24,12 @@ The following emojis are used to highlight certain changes:
+* 🛠 The `path` package has been massively refactored. With this refactor, we have
+  condensed the different path-related packages under a single one. Therefore, there
+  are many breaking changes. Please consult the [documentation](https://pkg.go.dev/github.com/ipfs/boxo/path)
+  for more details on how to use the new package.
```
I will add it, but for the go docs i think is fine: I already link to `path.CLean` from Go which includes that documentation. If anything changes there [very unlikely we wouldn't have to recopy/paste.

hacdias commented on path/resolver/resolver.go:
```diff
@@ -24,76 +22,73 @@ import (
-	if err != nil {
-		return cid.Cid{}, nil, err
-	}
+	c, remainder := fpath.Cid(), fpath.Segments()[2:]
```
3 places only. I think we can add it in the future if we feel the need for it.

hacdias commented:
> To avoid making waves this difficult month, I think we should minimize risk and ship this in Kubo 0.24, and we should wait with merge until 0.23 ships.

It already has: https://github.com/ipfs/kubo/releases/tag/v0.23.0!

hacdias commented:
I am going to go ahead and merge this. I think it has marinated for enough time and I've already addressed the feedbacks we had. Tests are passing here and in Kubo, and Kubo 0.23 has been released. So merging this at the beginning of the new release cycle (0.24) will allow us to test it until the final release of 0.24.
